Disallow concurrent consumer group migrations

bashtanov opened the pull request:
Currently data migration reconciliation mechanism in `backend` and `worker` classes assume each topic and each partition only belongs to no more than one migration.

When this is fixed we can lift this restriction.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
rptest/data_migrations_api_test: assure_not_migratable to support groups

also make expected_response non-optional

bashtanov committed:
c/data_migrated_table: disallow concurrent migrations of consumer groups

Currently data migration reconciliation mechanism in `backend` and
`worker` classes assume each topic and each partition only belongs to
no more than one migration.

When this is fixed we can lift this restriction.

bashtanov committed:
rptest/data_migrations_api_test: disallow concurrent group migrations

check concurrent group migrations cannot be created

joe-redpanda commented on tests/rptest/tests/data_migrations_api_test.py:
```diff
@@ -306,6 +315,7 @@ def test_creating_with_topic_no_remote_writes(self):
         )
         self.assure_not_migratable(
             topic,
+            None,
```
Nit: can you use named parameters for these

joe-redpanda commented on tests/rptest/tests/data_migrations_api_test.py:
```diff
@@ -283,6 +283,30 @@ def test_conflicting_migrations(self):
+            [make_namespaced_topic(topic.name)], consumer_groups=[]
+        )
+        self.create_and_wait(out1)
+        self.assure_not_migratable(
```
"test_conflicting_group_migrations" but we're not migrating groups here?

out1 involves no consumer groups

nor does the migration pitched to assure_not_migratable

two tests?
1. two on the same topic to show that the same topic can't be migrated twice at the same time
2. two different topics but they attempt to pull the same consumer group with

bashtanov replied:
:facepalm: I lost it in a rebase

joe-redpanda commented on tests/rptest/tests/data_migrations_api_test.py:
```diff
@@ -283,6 +283,28 @@ def test_conflicting_migrations(self):
+            topic=None,
+            group="group2",
+            expected_response={
+                "message": "Unexpected cluster error: Requested operation can not be executed as the resource is undergoing data migration",
```
commentary:
Probably not the best error message given that no user-visible resource is shared between the two migrations, but IMO should be fine given that the restriction will be lifted in the future

joe-redpanda replied:
actually is this going to fail ci?
`cannot migrate groups when there are active group migrations`
seems to be the error in the cpp

bashtanov replied:
This string is only going to be logged. The one returned to caller is generated by `src/v/cluster/errc.h` from `errc::resource_is_being_migrated`.

joe-redpanda commented on src/v/cluster/data_migrated_resources.h:
```diff
@@ -48,10 +48,12 @@ class migrated_resources {
         return _topics.contains(topic);
     }
-    /// Checks if consumer group  is already migrated
+    /// Checks if consumer group is already migrated
```
nit:
should this and the test commit be switched? I dont know that the test will pass without the cpp changes

bashtanov replied:
We don't aim to make every commit pass CI, we only aim for every commit to be buildable for to bisect.

bashtanov committed:
c/test/data_migrated_table: create only 1 migration with consumer groups

We temporarily disallowed concurrent migrations with consumer groups
even if these consumer groups are distinct. Adjust unittests accordingly

bashtanov commented:
last push is just a rebase till latest dev

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75870-->
<!--fe8c41fc-22b5-46ff-b71f-83c94220085a-->
<!--title-->
## Retry command for [Build#75870](https://buildkite.com/redpanda/redpanda/builds/75870)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_conflicting_group_migrations
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--0681beeb-5acf-4193-893d-cad8fd8f10c6-->
<!--TEST RESULTS-->
<!--2e6e1c14-dae3-4003-9ec2-b2e5117ddbda-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75870">build#75870</a></summary>


| test_class            | test_method                                  | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:----------------------|:---------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_conflicting_group_migrations            | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75870#019a60bc-5960-4a84-809a-bc60df091378 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_conflicting_group_migrations    |
| SegmentMsTest         | test_segment_rolling_with_retention_consumer | null             | integration | https://buildkite.com/redpanda/redpanda/builds/75870#019a60bc-595a-461f-af88-e7bf107da611 | FLAKY         | 19/21    | upstream reliability is '93.76218323586744'. current run reliability is '90.47619047619048'. drift is 3.28599 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75877">build#75877</a></summary>


| test_class                      | test_method                                  | test_arguments                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:--------------------------------|:---------------------------------------------|:--------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| SegmentMsTest                   | test_segment_rolling_with_retention_consumer | null                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75877#019a6126-9703-45c8-ad0c-2e7359513008 | FLAKY         | 19/21    | upstream reliability is '93.70229007633588'. current run reliability is '90.47619047619048'. drift is 3.2261 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test                     | {"cloud_storage_type": 1, "mixed_versions": true} | integration | https://buildkite.com/redpanda/redpanda/builds/75877#019a6126-9708-458d-8839-c4d6e5d182de | FLAKY         | 18/21    | upstream reliability is '99.07161803713528'. current run reliability is '85.71428571428571'. drift is 13.35733 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test   |
</details>
<!--/body-->

vbotbuildovich commented:
/backport v25.3.x

vbotbuildovich commented:
Branch name "v25.3.x" not found.

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/19235344783).
